### DISCUSSION TITLE
Binding an AvaloniaDictionary to a Listbox?

### DISCUSSION THREAD

**Discussion by tlitookilakin** [AvaloniaUI/Avalonia#11374] [Q&A] [answered]
Hello! I've been working on a project that stores unique string IDs, and I'm using a listbox so users can select an ID to manipulate the data attached to it. 

Previously I was using a combination of `AvaloniaDictionary`s, `AvaloniaList`s, and an internal ID property, but as you can imagine this became unwieldy quite quickly. I rewrote the system to using a single `AvaloniaDictionary`, but now I can't bind it to a ListBox (as it doesn't implement IList). My next attempt was to try and bind to the `Key` and `Value` collections separately, which didn't throw an error, but also wouldn't update when items were added or removed.

I've considered using an `AvaloniaList` of `KeyValuePair`s, but this removes the convenience of having a dictionary- whenever I wanted to retrieve a values or check if it exists, I would have to iterate the entire list.

What's the best way to go about doing this? I thought this would be a relatively common use-case but I haven't been able to find anyone else asking about this particular wrinkle.

**Comment by maxkatz6** [comment, marked as answer]
Looks like you want to implement a custom collection.
As your requirements are these:
1. It should work in ListBox with dynamic changes
2. It should be based on dictionary to allow quite add/remove
3. It needs to implement both IList and INotifyCollectionChanged (Avalonia requirement to get `#1`)

To achieve this you can create a simple collection, which implement these interfaces and also uses dictionary internally.
And you can check AvaloniaDictionary (and AvaloinaLlist) source code to see how they work internally: https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Base/Collections/AvaloniaDictionary.cs

Important details is that AvaloniaDictionary (as simple dictionary too) operates with KeyValueItem<TKey, TValue>, and not TValue directly. Meaning all change notification will report full KeyValueItem pair. 

If that's fine and you can use KeyValueItem in the UI (I assume you can, since you tried to use Dictionary as a source directly), you can create custom collection like this:

```
public class AvaloinaDictionaryList<TKey, TValue> : AvaloniaDictionary<TKey, TValue>, IList<KeyValuePair<TKey, TValue>>
{
}
```
And implement `IList<KeyValuePair<TKey, TValue>>` members according to your needs (most likely just redirect them).

> I thought this would be a relatively common use-case

Doesn't really sounds like one. But DynamicData from ReactiveUI does solve similar problems.

**Comment by tlitookilakin** [comment]
Thanks a bunch!
